server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -2082,12 +2082,59 @@ void handleBonusPools(Owner owner, Map<String, PoolQuantity> pools,
         @Override
         public void handlePostEntitlement(PoolManager manager, Consumer consumer,
             Map<String, Entitlement> entitlements) {
-            Set<String> stackIds = new HashSet<String>();
+
+            /*
+             * selectBestPools can return multiple pools on the same stack.
+             * Hence there can be more than one entitlement with the same stack id here.
+             * enforcer.postEntitlement may try to create more than one stack derived pool
+             * per stack id if we provide that as input, so we sort the entitlements.
+             *
+             * Entitlements that share stack ids need to be operated individually.
+             * other entitlements can be operated in a batch.
+             *
+             */
+            Map<String, Integer> stackIdCounts = new HashMap<String, Integer>();
             for (Entitlement entitlement : entitlements.values()) {
                 if (entitlement.getPool().isStacked()) {
-                    stackIds.add(entitlement.getPool().getStackId());
+                    String stackId = entitlement.getPool().getStackId();
+                    Integer count = stackIdCounts.containsKey(stackId) ?
+                        stackIdCounts.get(stackId) : 0;
+                    stackIdCounts.put(stackId, count + 1);
                 }
             }
+            Set<String> uniqueStackIds = new HashSet<String>();
+            Map<String, Entitlement> uniqueEnts = new HashMap<String, Entitlement>();
+            Map<String, Entitlement> nonUniqueEnts = new HashMap<String, Entitlement>();
+            for (Entry<String, Entitlement> entry : entitlements.entrySet()) {
+                Pool p = entry.getValue().getPool();
+                if (p.isStacked() && stackIdCounts.get(p.getStackId()) > 1) {
+                    nonUniqueEnts.put(entry.getKey(), entry.getValue());
+                }
+                else {
+                    uniqueEnts.put(entry.getKey(), entry.getValue());
+                    if (p.isStacked()) {
+                        uniqueStackIds.add(p.getStackId());
+                    }
+                }
+            }
+            // ents with unique stack id can be operated in a batch
+            handlePostEntitlement(manager, consumer, uniqueEnts, uniqueStackIds);
+            // others must be operated individually
+            for (Entry<String, Entitlement> entry: nonUniqueEnts.entrySet()) {
+                Entitlement ent = entry.getValue();
+                Map<String, Entitlement> entMap = new HashMap<String, Entitlement>();
+                entMap.put(entry.getKey(), ent);
+                Set<String> stackIdSet = new HashSet<String>();
+                if (ent.getPool().isStacked()) {
+                    stackIdSet.add(ent.getPool().getStackId());
+                }
+                handlePostEntitlement(manager, consumer, entMap, stackIdSet);
```

Comment:
Just seems like if we have a group of ents of the same stack, then we should be able to update the subpool in one call. I'm probably making this seem a little easier than it actually is.

Might be nice to have some tests around that filtering code, if it isn't already covered by other tests.